lib/Dialect/Verif/Transforms/VerifyClockedAssertLike.cpp

```diff
@@ -0,0 +1,101 @@
+//===- VerifyClockedAssertLike.cpp - Check Clocked Asserts -------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Checks that clocked assert-like ops are constructed correctly.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWModuleGraph.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/LTL/LTLOps.h"
+#include "circt/Dialect/LTL/LTLTypes.h"
+#include "circt/Dialect/Verif/VerifDialect.h"
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Dialect/Verif/VerifPasses.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace verif;
+
+namespace {
+// Verify function for clocked assert / assume / cover ops.
+// This checks that they do not contiain any nested clocks or disable operations
+// Clocked assertlike ops are a simple form of assertions that only
+// contain one clock and one disable condition.
+struct VerifyClockedAssertLikePass
+    : VerifyClockedAssertLikeBase<VerifyClockedAssertLikePass> {
+  void runOnOperation() override;
+
+private:
+  static void verify(Operation *clockedAssertLikeOp) {
+    // Used to perform a DFS search through the module to visit all operands
+    // before they are used
+    llvm::SmallMapVector<Operation *, OperandRange::iterator, 16> worklist;
+
+    // Keeps track of operations that have been visited
+    llvm::DenseSet<Operation *> handledOps;
```

Comment:
You may want to pull this `handledOps` set out into the pass itself (and clear it at the beginning of the pass). If I understand correctly, this is the set of operations that contain no nested `ltl.clock` and `ltl.disable` ops. If a subsequence is used in multiple asserts or assumes, which may be a common thing if CSE does its job well, remembering which ops you've already deemed safe avoids re-visiting those common ops over and over again.

For example:
```
%0 = ltl.delay ...
verif.clocked_assert %0
verif.clocked_assert %0
verif.clocked_assert %0
```
would cause you to traverse the entire expression tree three times if you keep `handledOps` scoped locally in `verify`; but if you preserve it throughout the run of the pass, you'll only check the `%0` expression tree once.